### PR TITLE
added floating window to indicate and stop recording

### PR DESCRIPTION
**Pull request by aryangupta701**
- added a floating window to indicate and stop the recording process

Screenshot from 2023-08-19 15-24-03

Screenshot from 2023-08-19 15-23-55

### CODE REVIEW

**Review by kingthorin — APPROVED** [review]
Looks good and seems reasonable to me.

**Comment by aryangupta701** [comment]
I will create a separate PR for reporting the recorded actions over this floating window

**Comment by thc202** [comment]
This should be rebased and use/extract constants. (It has conflicts too.)

**Comment by thc202** [comment]
It seems the events done to the floating window might also end up being recorded to the scripts, e.g.:
```json
    {
      "windowHandle": "windowHandle1",
      "type": "id",
      "element": "ZapfloatingDiv",
      "index": 4,
      "enabled": true,
      "elementType": "ZestClientElementClick"
    }
```

**Comment by aryangupta701** [comment]
ohhh Thank you. Forgot to test that.

File: source/ContentScript/recorder.ts [hunk 1]
```diff
@@ -206,15 +208,119 @@ class Recorder {
...
+    const floatingDiv = document.getElementById('ZapfloatingDiv');
+    if (floatingDiv) {
+      floatingDiv.style.display = 'none';
+    }
+  }
+
+  insertFloatingPopup(): void {
+    if (this.floatingWindowInserted) {
+      const floatingDiv = document.getElementById('ZapfloatingDiv');
+      console.log('hello');
```
Comment (thc202) [inline source/ContentScript/recorder.ts:240]: Probably better to remove these, or reworded to be more useful.

File: source/Popup/index.tsx [hunk 2]
```diff
@@ -183,7 +181,13 @@ function downloadZestScript(zestScriptJSON: string, title: string): void {
   closePopup();
 }
 
-function handleSaveScript(): void {
+async function handleSaveScript(): Promise<void> {
```
Comment (thc202) [inline source/Popup/index.tsx:184]: Better fix this in its own PR, it's unrelated to the changes being added here.

File: source/ContentScript/recorder.ts [hunk 3]
```diff
@@ -24,7 +24,7 @@ import {
   ZestStatementElementClick,
   ZestStatementElementSendKeys,
   ZestStatementLaunchBrowser,
-  ZestStatementSwichToFrame,
+  ZestStatementSwitchToFrame,
```
Comment (thc202) [inline source/ContentScript/recorder.ts:27]: This should be done in its own PR.

Comment (aryangupta701) [reply source/ContentScript/recorder.ts:27]: okay

File: source/Popup/index.tsx (hunk 2 above)
Comment (aryangupta701) [reply source/Popup/index.tsx:184]: okay

**Comment by aryangupta701** [comment]
ready for review

File: source/ContentScript/recorder.ts [hunk 4]
```diff
@@ -170,6 +172,12 @@ class Recorder {
     });
   }
 
+  shouldRecord(element: HTMLElement): boolean {
+    if (!this.active) return this.active;
+    if (element.style.zIndex === '999999') return false;
```
Comment (thc202) [inline source/ContentScript/recorder.ts:177]: The target page might be using this value as well, wouldn't it be better to check the ID of the elements?

File: source/ContentScript/recorder.ts [hunk 5]
```diff
@@ -193,15 +201,118 @@ class Recorder {
...
+    buttonElement.style.color = 'white';
+    buttonElement.style.zIndex = '999999';
+    buttonElement.style.border = 'none';
+    buttonElement.style.borderRadius = '3px';
+    buttonElement.style.cursor = 'pointer';
+    buttonElement.textContent = 'Stop Recording';
+
+    buttonElement.addEventListener('click', () => {
+      this.stopRecordingUserInteractions();
+      Browser.runtime.sendMessage({type: 'stopRecording'});
```
Comment (thc202) [inline source/ContentScript/recorder.ts:277]: Use the constant?

File: source/ContentScript/recorder.ts (hunk 4 above)
Comment (aryangupta701) [reply source/ContentScript/recorder.ts:177]: yeah would be better. Will fix this.

**Comment by thc202** [comment]
Sometimes the floating window is almost hidden, e.g.:
floating partially hidden

**Comment by aryangupta701** [comment]
fixing this

**Comment by aryangupta701** [comment]
if someone drag and drop to a particular location it will stay at that location even if the window size changes that may lead to the loss of the floating window. Thinking about how we can fix this

**Comment by aryangupta701** [comment]
Fixed :)

**Comment by thc202** [comment]
Looks better but it's still partially hidden:
partial

(That's the default window size when launching the browser from ZAP.)

**Comment by aryangupta701** [comment]
Fixing..

**Comment by aryangupta701** [comment]
I think this is again a mac related issue. For me it works fine. I can't reproduce this behavior @thc202

**Comment by thc202** [comment]
What information can I provide?

**Comment by aryangupta701** [comment]
I can suggest some hit and trials. Can you test them please? 
1. try changing https://github.com/zaproxy/browser-extension/pull/63/files#diff-326f362eacd2a9bea9debec8ede73c13ba68fdcbbb2c69a9f40a90d5afc3d5b0R246 replace -105 with -100

**Comment by aryangupta701** [comment]
I am able to reproduce this issue. I wonder why the chrome browser launched from ZAP and the the one which I directly use is rendering css differently.

**Comment by aryangupta701** [comment]
Ohh I got it. This is because of the website only. It has some css which is being applied over this floating window as well. So I guess I will need to use frames.

**Comment by aryangupta701** [comment]
ready for review

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by psiinon — APPROVED** [review]
This works well and gives the user really useful feedback - great work!
